File: 333-Programming-Languages/notes/sorter.cpp
#include <iostream>
#include <string>
using namespace std;

/**
 * Demonstrates defining a polymorphic sorting 
 * operation using templates for types that have
 * a < operator defined.
 */

template <class T>
void sort(T* a, int length)
{
	// sorts elements of a in place
	// at each iteration of the outer loop
	// the element at a[i] is properly ordered
	for (int i = 0; i < length; i++)
	{
		for (int j = i+1; j < length; j++)
		{
			if (a[j] < a[i])
			{
				T swap = a[j];
				a[j] = a[i];
				a[i] = swap;
			}
		}
	}
}

int main(int argc, char** argv)
{
	// convert command line arguments to 
	// an array of strings
	string array[argc];
	for (int i = 0; i < argc; i++)
	{
		array[i] = string(argv[i]);
		cout << array[i] << endl;
	}
	
	cout << "---" << endl;
	
	// creates a concrete sort routine for string
	sort<string>(array, argc);
	for (int i = 0; i < argc; i++)
	{
		cout << array[i] << endl;
	}
	
	// creates a concrete sort routine for char
	sort<char>(argv[argc-1], strlen(argv[argc-1]));
	cout << argv[argc-1] << endl;
}

File: 333-Programming-Languages/notes/Sorter.java
/**
 * Demonstrates a polymorphic sorting operation using
 * using interfaces for classes that implement the 
 * Comparable interface.
 */
public class Sorter
{
	public static void sort(Comparable[] a)
	{
		// sorts elements of a in place
		// at each iteration of the outer loop
		// the element at a[i] is properly ordered
		for (int i = 0; i < a.length; i++)
		{
			for (int j = i+1; j < a.length; j++)
			{
				if (a[j].compareTo(a[i]) < 0)
				{
					Comparable swap = a[j];
					a[j] = a[i];
					a[i] = swap;
				}
			}
		}
	}
	
	public static void main(String[] args)
	{
		for (String str : args)
			System.out.println(str);
		System.out.println("---");
		sort(args);
		for (String str : args)
			System.out.println(str);
	}
}